daemon/logger/journald/read.go

```diff
@@ -167,30 +169,82 @@ func (s *journald) Close() error {
 	return nil
 }
 
-func (s *journald) drainJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, oldCursor string) string {
+func (s *journald) drainJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, oldCursor string) (string, bool) {
 	var msg, data, cursor *C.char
 	var length C.size_t
 	var stamp C.uint64_t
 	var priority, partial C.int
 
+	// Give the journal handle an opportunity to close any open descriptors
+	// for files that have been removed.
+	C.sd_journal_process(j)
+
+	// Seek to the location of the last entry that we sent.
+	eof := false
+	if oldCursor != "" {
+		startCursor := C.CString(oldCursor)
+		defer C.free(unsafe.Pointer(startCursor))
+		// We know which entry was read last, so try to go to that
+		// location.
+		rc := C.sd_journal_seek_cursor(j, startCursor)
+		if rc != 0 {
+			return oldCursor, false
+		}
+		// Go forward to the first unsent message.
+		rc = C.sd_journal_next(j)
+		if rc < 0 {
+			return oldCursor, false
+		}
+		// We want to avoid sending a given entry twice (or more), so
+		// attempt to advance to the first unread entry in the journal
+		// so long as "this" one matches the last entry that we read.
+		for C.sd_journal_test_cursor(j, startCursor) > 0 {
+			if C.sd_journal_next(j) <= 0 {
+				return oldCursor, false
+			}
+		}
+	}
+
 	// Walk the journal from here forward until we run out of new entries.
-drain:
+	var sent uint64
 	for {
-		// Try not to send a given entry twice.
-		if oldCursor != "" {
-			ccursor := C.CString(oldCursor)
-			defer C.free(unsafe.Pointer(ccursor))
-			for C.sd_journal_test_cursor(j, ccursor) > 0 {
-				if C.sd_journal_next(j) <= 0 {
-					break drain
-				}
+		// If we're not keeping up with journald writing to the journal, some of the
+		// files between where we are and "now" may have been deleted since we started
+		// walking the set of entries.  If that's happened, the inotify descriptor in
+		// the journal handle will have pending deletion events after we've been reading
+		// for a while.  Letting the journal library process them will close any that
+		// are already deleted, so that we'll skip over them and allow space that would
+		// have been reclaimed by deleting these files to actually be reclaimed.
+		if sent > 0 && sent%1024 == 0 {
+			if status := C.sd_journal_process(j); status < 0 {
+				cerrstr := C.strerror(C.int(-status))
+				errstr := C.GoString(cerrstr)
+				fmtstr := "error %q while attempting to process journal events for container %q"
+				logrus.Errorf(fmtstr, errstr, s.vars["CONTAINER_ID_FULL"])
+				// Attempt to rewind the last-read cursor to the
+				// entry that we last sent.
+				C.sd_journal_previous(j)
```

Comment:
Seems like you might need to test for success here.  If it fails, given the error above, probably worth just bailing entirely.

Reply:
Is it then also worth to check the errors from all calls to `sd_journal_next()`/`sd_journal_previous()`?

Reply:
My concern with checking the result of `sd_journal_previous()` is that it will fail if we happen to be on the first matching entry in the first file in the journal, which could happen when files containing previous entries have just been deleted.  In those cases I'd lean more in the direction of mistakenly skipping an entry than failing the whole read operation.  We should be checking for errors from `sd_journal_next()` consistently, since that function failing usually means we've reached the end of the journal.

Reply:
But what about checking for strictly less than 0?  That usually means a problem.

Reply:
There's no harm in checking for it and logging the error, and I've added some logic to do that, but there's not much else to be done at that point, since we're already breaking out of the reading loop.